LCLSGeneral/LCLSGeneral/POUs/Data/FB_BasicStats.TcPOU

```diff
@@ -0,0 +1,124 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1" ProductVersion="3.1.4022.18">
+  <POU Name="FB_BasicStats" Id="{dd720d4e-bff0-4eac-9bd3-db3174a49483}" SpecialFunc="None">
+    <Declaration><![CDATA[FUNCTION_BLOCK FB_BasicStats
+(*
+	Minimalist Array Stats for LREALs
+	2019-10-10 Zachary Lentz
+	
+	Calculates the most basic stats for an array and provides pytmc control points.
+	This is an alternative to the TC3 Condition Monitoring library which requires an
+	additional license and had a more complicated interface.
+*)
+VAR_IN_OUT
+	// Input array of floats
+	{attribute 'pytmc' := '
+		pv: STATS:DATA
+		io: i
+	'}
+	aSignal: ARRAY[*] OF LREAL;
+END_VAR
+VAR_INPUT
+	// If TRUE, we will update the outputs based on aSignal
+	bEnable: BOOL;
+	// If nonzero, we will only pay attention to the first nElems items in aSignal
+	{attribute 'pytmc' := '
+		pv: STATS:NELM
+		io: i
+	'}
+	nElems: UDINT;
+END_VAR
+VAR_OUTPUT
+	// Average of all values in the array
+	{attribute 'pytmc' := '
+		pv: STATS:MEAN
+		io: i
+	'}
+	fMean: LREAL;
+	// Standard deviation of all values in the array
+	{attribute 'pytmc' := '
+		pv: STATS:STDEV
+		io: i
+	'}
+	fStDev: LREAL;
+	// Root mean square of all values in the array
+	{attribute 'pytmc' := '
+		pv: STATS:RMS
+		io: i
+	'}
+	fRMS: LREAL;
+	// Largest value in the array
+	{attribute 'pytmc' := '
+		pv: STATS:MAX
+		io: i
+	'}
+	fMax: LREAL;
+	// Smallest value in the array
+	{attribute 'pytmc' := '
+		pv: STATS:MIN
+		io: i
+	'}
+	fMin: LREAL;
+	// Largest array element subtracted by the smallest
+	{attribute 'pytmc' := '
+		pv: STATS:RANGE
+		io: i
+	'}
+	fRange: LREAL;
+END_VAR
+VAR
+	nIndex: DINT;
+	nElemsSeen: UDINT;
+	fSum: LREAL;
+	fSquareSum: LREAL;
+	fSquareMean: LREAL;
+	fVarianceSum: LREAL;
+	fVarianceMean: LREAL;
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[IF bEnable THEN
+	// First pass through aSignal, get sum, square sum, mean, rms, max, min
+	nElemsSeen := 0;
+	fSum := 0;
+	fSquareSum := 0;
+	fMax := aSignal[LOWER_BOUND(aSignal, 1)];
+	fMin := fMax;
+	FOR nIndex := LOWER_BOUND(aSignal, 1) TO UPPER_BOUND(aSignal, 1) DO
+		nElemsSeen := nElemsSeen + 1;
+		fSum := fSum + aSignal[nIndex];
+		fSquareSum := fSquareSum + aSignal[nIndex] * aSignal[nIndex];
```

Comment:
LREAL (64-bit float) is unlikely to overflow here, though a loss of precision could be a possibility.

Reply:
Is there a better way to approach this? While writing the code I considered the overflow but neglected the precision. I suppose the alternative is to divide as I go? Which way has the milder precision hit?

Reply:
I seem to recall that there are some good algorithms for running calculations like this, but google is failing me...

For what it's worth, I think this is plenty good for now!

Reply:
Also, I'm considering removing the RMS entirely or replacing it with RMSE because it's pretty useless as a statistic without a reference point.

Reply:
A simple reference point default could just be the initial value - i.e., when the user last reset it. Then a `pytmc` reference value would be nice, too. I'd be OK with either adding onto this or removing it altogether. We can always buffer + calculate what we please in the IOC/Python, after all.

Reply:
OK, I'm opting to scrap this. We can add it back once we know what we want.

Reply:
Note that the stdev is almost just the RMS relative to the mean